internal/dag/gatewayapi_processor.go

```diff
@@ -745,13 +746,58 @@ func (p *GatewayAPIProcessor) validateBackendRef(backendRef gatewayapi_v1alpha2.
 		return nil, fmt.Errorf("Spec.Rules.BackendRef.Port must be specified")
 	}
 
+	// If the backend is in a different namespace than the route, then we need to
+	// check for a ReferencePolicy that allows the reference.
 	if backendRef.Namespace != nil && string(*backendRef.Namespace) != routeNamespace {
-		// we haven't implemented ReferencePolicy yet, so disallow any
-		// cross-namespace reference.
-		return nil, fmt.Errorf("Spec.Rules.BackendRef.Namespace must match the route's namespace")
+		var allowed bool
+		for _, referencePolicy := range p.source.referencepolicies {
+			// The ReferencePolicy must be defined in the namespace of
+			// the "referent" (i.e. the Service).
+			if referencePolicy.Namespace != string(*backendRef.Namespace) {
+				continue
+			}
+
+			// "From" must contain an entry matching the route that is
+			// referencing the service.
+			var fromAllowed bool
+			for _, from := range referencePolicy.Spec.From {
+				if from.Namespace == gatewayapi_v1alpha2.Namespace(routeNamespace) && from.Group == gatewayapi_v1alpha2.GroupName && from.Kind == gatewayapi_v1alpha2.Kind(routeKind) {
+					fromAllowed = true
+					break
+				}
+			}
+			if !fromAllowed {
+				continue
+			}
+
+			// "To" must contain an entry matching the Service that
+			// is being referenced.
+			var toAllowed bool
+			for _, to := range referencePolicy.Spec.To {
+				if (to.Group == "" || to.Group == "core") && to.Kind == "Service" && (to.Name == nil || to.Name == gatewayapi.ObjectNamePtr("") || to.Name == &backendRef.Name) {
```

Comment:
nit: only test cases I think might be missing are:
- when the reference policy has a name in the `To` section
- when there is a reference policy that would possible be valid but not in the right namespace

(just to get these pesky warnings about line coverage out of here as well)

Reply:
good call, got a little lazy here, also caught a pointer bug writing the additional test cases

Reply:
gotta make some more fixes to that last commit still.

Reply:
OK should be thoroughly covered now, thanks for calling this out.